File: tmp/qut_task_bst/bst.cpp
/*
 *     File Name: bst.c
 *     Author: Bo Yu
 *     Mail: tsu.yubo@gmail.com
 *     Created Time: 2017年12月21日 星期四 19时43分55秒
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>

#include <iostream>
using namespace std;

#define DBL_MAX 1.7976931348623158e+308
int n = 5; /* 5个结点*/

double OPTIMAL_BST(double *p, double *q, int n, double **e, double **w, int **root)
{
	int i, j, l,r;
	for(i = 1; i <= n + 1; i++){
		e[i][i-1] = q[i-1];
		w[i][i-1] = q[i-1];
	}
	for (l = 1; l <= n; l++){
		for(i=1; i<=n-l+1; i++){
			int j=i+l-1;
			e[i][j] = DBL_MAX;
			w[i][j] = w[i][j-1]+p[j] +q[j];
			for(r = i; r <= j; r++){
				double t = e[i][r-1]+e[r+1][j] + w[i][j];
				if(t < e[i][j]){
					e[i][j] = t;
					root[i][j] = r;
				}
			}
		}
	}
	return e[1][n];
}
void construct_optimal_bst(int **root, int i, int j)
{
	if (i == 1 && j == n)
		printf("K%d为根\n", root[i][j]);
	if (i < j){
		printf("K%d是K%d的左孩子\n", root[i][root[i][j]-1], root[i][j]);
		construct_optimal_bst(root, i, root[i][j] - 1);
		if (root[i][j] + 1 < j)
			printf("K%d是K%d的右孩子\n", root[root[i][j]+1][j], root[i][j]);
		construct_optimal_bst(root, root[i][j] +1, j);
	}
	if (i ==j){
		printf("d%d是K%d的左孩子\n", i-1, i);
		printf("d%d是K%d的右孩子\n", i, i);
	}
}
int main()
{
	double p[6] = {0, 0.15, 0.10, 0.05, 0.10, 0.20};
	double q[6] = {0.05, 0.10, 0.05, 0.05, 0.05, 0.10};
	/* 下标范围：e[1..n+1,0..n]  w[1..n+1,0..n] root[1..n,1..n] */
	double **e=new double *[n+2];
	for(int i=0;i<=n+1;i++) e[i]=new double[n+1];
	double **w=new double *[n+2];
	for(int i=0;i<=n+1;i++) w[i]=new double[n+1];
	int **root=new int *[n+1];
	for(int i=0;i<=n;i++) root[i]=new int[n+1];
        /* 计算表e和root */
        cout<<"最优期望搜索代价："<<OPTIMAL_BST(p,q,n,e,w,root)<<endl;
        /* 打印root */
	 construct_optimal_bst(root,1,n);
	 /* 回收内存 */
	 for (int i = 0; i < n+2; i++)
	{
		 delete e[i],delete w[i];
		e[i] = NULL,w[i] = NULL;
		 if(i!=n+1)
		 {
			delete root[i];
			root[i] = NULL;
		}
	}
			delete []e,delete[]w,delete []root;
			e = NULL,w=NULL,root=NULL;
    return 0;

}

File: tmp/qut_task_bst/bst.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-
# File Name: bst.py
# Author: Bo Yu
# mail: tsu.yubo@gmail.com
# Created Time: 2017年12月22日 星期五 09时25分39秒

import sys, time

gk = lambda i,j:str(i)+',' + str(j)

# p 是关键字的概率(n个)，ｑ是伪关键字的概率，ｎ是关键字的个数
# 设置最大值，取出p和q中的最大的一个概率值，乘以n+1(深度)，　共有
# len(p) + len(q)个结点，
def optimal_BST(p, q, n):
    MAX = (max(p) > max(q) and max(p) or max(q))*(n + 1)*(len(p)+len(q))


#定义三个集合
#e存放期望，w存放概率和，root存放子树的根节点
    e, w, root = {}, {}, {}
    for i in xrange(1, n + 2):
        e[gk(i, i - 1)] = q[i - 1]
        w[gk(i, i - 1)] = q[i - 1]
    for l in xrange(1, n + 1):
        for i in xrange(1, n - l + 2):
            j = i + l - 1
            e[gk(i, j)] = MAX
            w[gk(i, j)] = w[gk(i, j - 1)] + p[j - 1] + q[j]
            for r in xrange(i, j + 1):
                t = e[gk(i, r - 1)] + e[gk(r+1,j)] + w[gk(i, j)]
                if t < e[gk(i, j)]:
                    e[gk(i, j)] = t
                    root[(gk(i,j))] = r
    return e, root

# p 是关键字的概率(n个)，ｑ是伪关键字的概率，ｎ是关键字的个数
def main():
    p = [0.15, 0.10, 0.05, 0.10, 0.20]
    q = [0.05, 0.10, 0.05, 0.05, 0.05, 0.10]
    e, root = optimal_BST(p, q, len(p))
    print e[gk(1, len(p))]

if __name__ == '__main__':
    main()
